src/adapter/src/optimize/peek.rs

```diff
@@ -316,8 +314,30 @@ impl<'s> Optimize<LocalMirPlan<Resolved<'s>>> for Optimizer {
             &self.typecheck_ctx,
             &mut df_meta,
         );
+
+        // Let's already try creating a fast path plan. If successful, we don't need to run the
+        // whole optimizer pipeline, but just a tiny subset of it. (But we'll need to run
+        // `create_fast_path_plan` later again, because, e.g., running `LiteralConstraints` is still
+        // ahead of us.)
+        let use_fast_path_optimizer = match create_fast_path_plan(
```

Comment:
It might be a bit wasteful to create the fast path plan here and then run through the optimization pipeline again? Especially when we already here determine it's a constant or simple peek, we could probably use the result instead of running the optimizer at all? I'm fine with merging as-is, but we should take note.

Reply:
> run through the optimization pipeline again

Do you mean "again", in the sense that we have already run the `optimize_mir_local` optimizer [above](https://github.com/MaterializeInc/materialize/blob/72a84836e8505131008d68353c9a8db882350a09/src/adapter/src/optimize/peek.rs#L186)? Running `optimize_dataflow` below still does meaningful things, i.e., not all the things that we want to happen for fast path peeks are done in `optimize_mir_local`. The most important thing that we still need to do is `LiteralConstraints`, which can turn a full index scan into an index lookup. (And then we need `CanonicalizeMfp` again, because `LiteralConstraints` can leave the MFP in a non-canonical form. And we also need `FoldConstants` again, because `LiteralConstraints` might have got us a constant, if it found contradicting literal constraints. And `ReduceScalars` is just good to run all the time for various obscure reasons, see the last paragraph of 3. [here](https://materializeinc.slack.com/archives/C0761MZ3QD9/p1720117362807379).)

(Note that `create_fast_path_plan` doesn't run any optimizations. It should be a cheaper things than even the `fast_path_optimizer`.)

A possible improvement that we could still do here is to not even run the `optimize_mir_local` for peeks that look like they are gonna be fast path, but just run `fast_path_optimizer` on the initial MIR plan. I tried to do this when originally working on this PR, and ran into some minor troubles, which I can't recall at the moment (maybe it was just that it would have required a somewhat larger refactoring of sequencing), so I just decided to keep this PR simple. But if the time comes when the optimizer run time becomes a significant portion of the run time of fast path peeks again, due to optimizations in other parts of the system, then I'm happy to revisit this. (I think I did some measurements with just hackily commenting out `optimize_mir_local`, and it showed that the run time of `optimize_mir_local` is significantly lower than the full optimizer pipeline, which this PR is eliminating.)

Reply:
I've created an issue for eliminating `optimize_mir_local` from fast path peeks, so that this doesn't get forgotten: https://github.com/MaterializeInc/materialize/issues/29501